### PR TITLE
[all] Update k8s dependencies to v1.19.0

### PR DESCRIPTION
**Pull request by seanschneeweiss**
**What this PR does / why we need it**:
The current [v1.18.0 service controller](https://github.com/kubernetes/kubernetes/blob/v1.18.0/pkg/controller/service/controller.go) for OCCM is missing two important fixes:
* [Service load balancers should include unschedulable nodes](https://github.com/kubernetes/kubernetes/commit/1f010cc42bb1e26d20e6c9729c78c9820f5519b2#diff-1346c951c0bb4e123323d28318609d2b)
* [service-controller-fixup](https://github.com/kubernetes/kubernetes/commit/977aeab3e542cb1725fd0448878d74e07c2546b2#diff-ebecfaaeccbe95596653e693dc9351e8)

Also the cloud-provider controllers were moved to [k8s.io/cloud-provider/controllers](https://github.com/kubernetes/cloud-provider/tree/master/controllers).

Therefore, an update/vendoring is suggested to [version v1.19.0](https://github.com/kubernetes/cloud-provider/tree/v0.19.0). 

**Which issue this PR fixes(if applicable)**:
fixes #

**Special notes for reviewers**:

**Release note**:

```release-note
Update k8s dependencies to v1.19.0
```

<sub>Sean Schneeweiss <sean.schneeweiss@daimler.com>, Daimler TSS GmbH, [legal info/Impressum](https://github.com/Daimler/daimler-foss/blob/master/LEGAL_IMPRINT.md)</sub>

### CODE REVIEW

**Comment by seanschneeweiss** [comment]
Some tests are failing - seems to be the error fixed in this [openlab PR](https://github.com/theopenlab/openlab-zuul-jobs/pull/948).
> `etcd version 3.4.9 or greater required.`

/test all

**Comment by ramineni** [comment]
@seanschneeweiss I did a recheck on jobs , could you check the failures now, if related to PR.

**Comment by seanschneeweiss** [comment]
@ramineni Thank you. Could you also check etcd version for _openlab/cloud-provider-openstack-acceptance-test-keystone-authentication-authorization_?
`etcd version 3.4.9 or greater required.`

I will check the errors on the other two jobs.

**Comment by seanschneeweiss** [comment]
Waiting for openlab-zuul-jobs [PR Update default etcd server to 3.4.9 #972](https://github.com/theopenlab/openlab-zuul-jobs/pull/972)

**Comment by wangxiyuan** [comment]
recheck

**Comment by ramineni** [comment]
@seanschneeweiss 1.19 is released, can you update go.mod to reflect that , we could debug on the final version if its failing?

**Comment by seanschneeweiss** [comment]
@ramineni yes sure, will probably do it on Friday. Thank you.

**Comment by jichenjc** [comment]
@seanschneeweiss k8s 1.19 is out, should we update to that instead of rc.4?

**Comment by seanschneeweiss** [comment]
@ramineni @jichenjc Could help debugging the test failures? I've upgraded to v1.19.0. Thank you.

**Comment by jichenjc** [comment]
```
2020-08-28 13:34:32.800225 | ubuntu-xenial-citynetwork | secret/cloud-config created
2020-08-28 13:34:32.808812 | ubuntu-xenial-citynetwork | + sleep 5
2020-08-28 13:34:37.811589 | ubuntu-xenial-citynetwork | + timeout 300 bash -c '
2020-08-28 13:34:37.811746 | ubuntu-xenial-citynetwork |     while :
2020-08-28 13:34:37.811779 | ubuntu-xenial-citynetwork |     do
2020-08-28 13:34:37.811904 | ubuntu-xenial-citynetwork |         /home/zuul/src/k8s.io/kubernetes/cluster/kubectl.sh get pod --all-namespaces | sed "1d" | awk '\''$4 != "Running" {err = 1} END {exit err}'\'' && break
2020-08-28 13:34:37.811946 | ubuntu-xenial-citynetwork |         sleep 1
2020-08-28 13:34:37.811989 | ubuntu-xenial-citynetwork |     done
2020-08-28 13:34:37.812012 | ubuntu-xenial-citynetwork |     '
2020-08-28 13:39:37.839620 | ubuntu-xenial-citynetwork | + echo 'Run services failed'
2020-08-28 13:39:37.840169 | ubuntu-xenial-citynetwork | Run services failed
2020-08-28 13:39:37.840261 | ubuntu-xenial-citynetwork | + /home/zuul/src/k8s.io/kubernetes/cluster/kubectl.sh get pod --all-namespaces
2020-08-28 13:39:38.031627 | ubuntu-xenial-citynetwork | NAMESPACE     NAME                            READY   STATUS    RESTARTS   AGE
2020-08-28 13:39:38.031796 | ubuntu-xenial-citynetwork | kube-system   csi-cinder-controllerplugin-0   0/5     Pending   0          5m6s
2020-08-28 13:39:38.036577 | ubuntu-xenial-citynetwork | + exit 1
2020-08-28 13:39:38.284558 | ubuntu-xenial-citynetwork | ERROR
2020-08-28 13:39:38.284971 | ubuntu-xenial-citynetwork | {
```

I saw this ,seems the pod is in pending ,not sure what happened ...maybe cpu/memory not enough in openlab test env?

**Comment by ramineni** [comment]
> @ramineni @jichenjc Could help debugging the test failures? I've upgraded to v1.19.0. Thank you.

@seanschneeweiss looks like there is a failure in initializing node , so taint is not removed, hence making pod not schedulable https://logs.openlabtesting.org/logs/51/1151/ceccf726ca5c1e91ef5405b6167716619783a6ee/cloud-provider-openstack-acceptance-test-csi-cinder/cloud-provider-openstack-acceptance-test-csi-cinder/bff2aa3/logs/kubernetes/cloud-controller-manager.log , Need to check whats changed in new version causing this error

**Comment by lingxiankong** [comment]
```
E0828 13:33:53.800027   12169 openstack.go:889] Failed to create an OpenStack LoadBalancer client: failed to find load-balancer v2 endpoint for region Lon1: No suitable endpoint could be found in the service catalog.
E0828 13:33:53.800041   12169 core.go:97] Failed to start service controller: the cloud provider does not support external load balancers
```

It seems that LB service is not deployed in Lon1 region of citynetwork cloud, I've already asked openlab maintainers, waiting for the answer.

**Comment by lingxiankong** [comment]
> ```
> E0828 13:33:53.800027   12169 openstack.go:889] Failed to create an OpenStack LoadBalancer client: failed to find load-balancer v2 endpoint for region Lon1: No suitable endpoint could be found in the service catalog.
> E0828 13:33:53.800041   12169 core.go:97] Failed to start service controller: the cloud provider does not support external load balancers
> ```
> 
> It seems that LB service is not deployed in Lon1 region of citynetwork cloud, I've already asked openlab maintainers, waiting for the answer.

Sorry, the above is not related, the service controller just skipped. The actual ERROR is:

```
E0828 13:34:00.636656   12169 node_controller.go:370] failed to initialize node ubuntu-xenial-citynetwork-citynetwork-openlab-0000145532 at cloudprovider: ProviderID "openstack://ubuntu-xenial-citynetwork-citynetwork-openlab-0000145532" didn't match expected format "openstack:///InstanceID"
```

The node failed to initialize, which led to pod creation PENDING.

**Comment by ramineni** [comment]
@seanschneeweiss I think I got the problem , recent changes in cloudprovider are passing node.spec.ProviderID directly, so we dont need 3 slashes here in regex, two should suffice , https://github.com/kubernetes/cloud-provider-openstack/blob/master/pkg/cloudprovider/providers/openstack/openstack_instances.go#L232 , could you make the change and check ?

File: pkg/cloudprovider/providers/openstack/openstack_instances.go [hunk 1]
```diff
@@ -46,6 +46,15 @@ const (
 
 // Instances returns an implementation of Instances for OpenStack.
 func (os *OpenStack) Instances() (cloudprovider.Instances, bool) {
+	return os.instances()
```
Comment (ramineni) [inline pkg/cloudprovider/providers/openstack/openstack_instances.go:49]: Is Instances() function still required when InstancesV2 implemented?

File: pkg/cloudprovider/providers/openstack/openstack_instances.go [hunk 2]
```diff
@@ -139,7 +153,13 @@ func (i *Instances) InstanceExistsByProviderID(ctx context.Context, providerID s
 	return true, nil
 }
 
-// InstanceShutdownByProviderID returns true if the instances is in safe state to detach volumes
+// InstanceShutdown returns true if the instance is shutdown.
+// It is the only state, where volumes can be detached immediately .
```
Comment (jichenjc) [inline pkg/cloudprovider/providers/openstack/openstack_instances.go:157]: maybe sync the comment for `InstanceShutdown ` and `InstanceShutdownByProviderID `

**Comment by jichenjc** [comment]
> 
> 
> @seanschneeweiss I think I got the problem , recent changes in cloudprovider are passing node.spec.ProviderID directly, so we dont need 3 slashes here in regex, two should suffice , https://github.com/kubernetes/cloud-provider-openstack/blob/master/pkg/cloudprovider/providers/openstack/openstack_instances.go#L232 , could you make the change and check ?

um... if you take a look
https://github.com/kubernetes/cloud-provider/blob/master/cloud.go#L100
it's last change is 2018 .. 

I knew the node as input param but it should not break the contract ?

**Comment by ramineni** [comment]
> > @seanschneeweiss I think I got the problem , recent changes in cloudprovider are passing node.spec.ProviderID directly, so we dont need 3 slashes here in regex, two should suffice , https://github.com/kubernetes/cloud-provider-openstack/blob/master/pkg/cloudprovider/providers/openstack/openstack_instances.go#L232 , could you make the change and check ?
> 
> um... if you take a look
> https://github.com/kubernetes/cloud-provider/blob/master/cloud.go#L100
> it's last change is 2018 ..
> 
> I knew the node as input param but it should not break the contract ?

@jichenjc hmm, may be I just based it on logs ,that there are no more 3 slashes from the o/p, need to check exact cause for that
E0828 13:34:00.636656   12169 node_controller.go:370] failed to initialize node ubuntu-xenial-citynetwork-citynetwork-openlab-0000145532 at cloudprovider: ProviderID "openstack://ubuntu-xenial-citynetwork-citynetwork-openlab-0000145532" didn't match expected format "openstack:///InstanceID"

**Comment by seanschneeweiss** [comment]
> @jichenjc hmm, may be I just based it on logs ,that there are no more 3 slashes from the o/p, need to check exact cause for that
> `E0828 13:34:00.636656 12169 node_controller.go:370] failed to initialize node ubuntu-xenial-citynetwork-citynetwork-openlab-0000145532 at cloudprovider: ProviderID "openstack://ubuntu-xenial-citynetwork-citynetwork-openlab-0000145532" didn't match expected format "openstack:///InstanceID"`

@jichenjc @ramineni Thank you, I appreciate your help. I'm still testing and investigating - so far, on our environment, we only receive an providerID in this format `openstack:///InstanceID`. Unsure where the changed providerID comes from...

I will push a small change soon, as `InstanceMetadataByProviderID` was renamed to `InstanceMetadata` from v1.19.0-rc.4 to v1.19.0.

**Comment by jichenjc** [comment]
take a look at 
https://logs.openlabtesting.org/logs/75/1175/4c8baa40b6d7f5d24ea1ce16025f4984f2a6db68/cloud-provider-openstack-e2e-test-csi-cinder/cloud-provider-openstack-e2e-test-csi-cinder/1115e92/logs/kubernetes/cloud-controller-manager.log

which is from another PR, even though there are simliar logs `ProviderID "openstack://ubuntu-xenial-citynetwork-citynetwork-openlab-0000146249" didn't match expected format "openstack:///InstanceID"` , the test still running (though failed but much different to this PR)

so My proposal is 
1) open a separate PR to fix the didn't match expected format "openstack:///InstanceID"` issue
2) add additional logs to check why the pods are pending like
```
2020-09-04 08:44:07.785040 | ubuntu-xenial-citynetwork | + /home/zuul/src/k8s.io/kubernetes/cluster/kubectl.sh get pod --all-namespaces
2020-09-04 08:44:07.984074 | ubuntu-xenial-citynetwork | NAMESPACE     NAME                            READY   STATUS    RESTARTS   AGE
2020-09-04 08:44:07.984201 | ubuntu-xenial-citynetwork | kube-system   csi-cinder-controllerplugin-0   0/5     Pending   0          5m5s
```
for 5 minutes 
3) I didn't find pods and other logs for further debug (such as why the pod is pending) 
https://logs.openlabtesting.org/logs/27/627/f0c6986fbb05fc3f4271331ef62487700b88abc1/check/cluster-api-provider-openstack-current-acceptance-test-v1.18/0205e80/logs/
from cluster-api-provider-openstack actually has all those info , we need consider to add it

**Comment by lingxiankong** [comment]
Just cinder related jobs were failed.

**Comment by jichenjc** [comment]
last failure still realted to cinder, the issues for all pods of CSI is pending which need further check

**Comment by jichenjc** [comment]
https://logs.openlabtesting.org/logs/51/1151/3c7ac3edc75f2ee7814d209699621eb31b694366/cloud-provider-openstack-acceptance-test-csi-cinder/cloud-provider-openstack-acceptance-test-csi-cinder/994aa25/job-output.txt.gz

has exact same errors ..

**Comment by jichenjc** [comment]
https://github.com/theopenlab/openlab-zuul-jobs/pull/1019 to debug this ...

**Comment by jichenjc** [comment]
@seanschneeweiss 
FYI  https://github.com/kubernetes/cloud-provider-openstack/pull/1187 (I used to use it to debug for `///` and I rebased this PR as well) the cinder related test cases passed and I think the root cause is InstanceV2 support

I disable the InstanceV2 (by return false) instead of true and nothing else , I guess it's not right timing to provide support for it without thorough test ... maybe you can check it as well? 

@lingxiankong @zetaab  @ramineni FYI

**Comment by lingxiankong** [comment]
I submitted a new PR (but @seanschneeweiss's commits retained) #1189, changed some versions in go.mod and disabled the `InstanceV2()` let's see the result.

**Comment by jichenjc** [comment]
@lingxiankong we should keep both PRs ? I guess #1189 first then this one?

**Comment by lingxiankong** [comment]
> @lingxiankong we should keep both PRs ? I guess #1189 first then this one?

No, I see @seanschneeweiss already submitted new commit to solve the issue, let's go with this one.

**Comment by seanschneeweiss** [comment]
Thank you everyone for all the support. I agree that the use of `InstancesV2` requires further analysis and improvement.

I've restarted `cloud-provider-openstack-acceptance-test-csi-manila` because it is stuck for over 6 hours now.
```
2020-09-07 14:04:23.514727 | ubuntu-bionic | pod/new-nfs-share-snap-restore-pod created
2020-09-07 14:04:23.522657 | ubuntu-bionic | + /home/zuul/src/k8s.io/kubernetes/cluster/kubectl.sh wait --for=condition=Ready pod/new-nfs-share-snap-restore-pod --timeout=300s
2020-09-07 14:09:23.962791 | ubuntu-bionic | error: timed out waiting for the condition on pods/new-nfs-share-snap-restore-pod
2020-09-07 14:09:23.966009 | ubuntu-bionic | + dump_container_logs
2020-09-07 14:09:23.966207 | ubuntu-bionic | + /home/zuul/src/k8s.io/kubernetes/cluster/kubectl.sh describe pods
2020-09-07 14:09:24.357930 | ubuntu-bionic | + /home/zuul/src/k8s.io/kubernetes/cluster/kubectl.sh describe pvc
2020-09-07 14:09:24.669518 | ubuntu-bionic | + /home/zuul/src/k8s.io/kubernetes/cluster/kubectl.sh describe pv
2020-09-07 14:09:24.938514 | ubuntu-bionic | + /home/zuul/src/k8s.io/kubernetes/cluster/kubectl.sh describe volumesnapshots
2020-09-07 14:09:25.195415 | ubuntu-bionic | + /home/zuul/src/k8s.io/kubernetes/cluster/kubectl.sh describe volumesnapshotcontents
2020-09-07 14:09:25.417181 | ubuntu-bionic | No resources found in default namespace.
```
and then:
```
TASK [Clean up resources for csi manila acceptance tests]
.....
+ /home/zuul/src/k8s.io/kubernetes/cluster/kubectl.sh delete -f examples/manila-csi-plugin/nfs/dynamic-provisioning/pvc.yaml
persistentvolumeclaim "new-nfs-share-pvc" deleted
```
after that it is stuck.

**Comment by lingxiankong** [comment]
Already suggested the openlab maintainers kill this job and restart.

**Comment by wangxiyuan** [comment]
> Already suggested the openlab maintainers kill this job and restart.

Done

**Comment by jichenjc** [comment]
/lgtm

looks like the manila CI still have some problem ... wait for that complete or maybe need trigger again

**Comment by lingxiankong** [comment]
@gman0 @gouthampacha @tombarron Could you please help to look at the failure of the job cloud-provider-openstack-acceptance-test-csi-manila?

**Comment by gman0** [comment]
@lingxiankong I'll have a look.

[csi-snapshotter logs](https://logs.openlabtesting.org/logs/51/1151/c300b97dcbc9f68842d7c51f9aedcda6f3811525/cloud-provider-openstack-acceptance-test-csi-manila/cloud-provider-openstack-acceptance-test-csi-manila/339c57a/logs/kubernetes/csi-snapshotter.txt):
```
E0908 07:43:44.636902       1 snapshot_controller.go:370] createSnapshot [create-default/new-nfs-share-snap[81b4c0af-5cb3-40c8-8d9b-1952f4fc51a2]]: error occurred in createSnapshotOperation: selfLink was empty, can't make reference
```

The snapshot is never created. This in turn fails to create a PVC from this snapshot. `pods/new-nfs-share-snap-restore-pod` mounts this PVC, but can't.

**Comment by gman0** [comment]
@ramineni and other csi-cinder folks: can you please confirm snapshotting works for you with k8s 1.19 and csi-snapshotter v1.2.2?

**Comment by jichenjc** [comment]
I tested locally #974 before the code merge (volume create ==> snapshot create )
but not with very thorough test due to no CI

why cinder CSI snapshot is related to manila ?

**Comment by gman0** [comment]
@jichenjc Ah, I missed #974. Both CSI Cinder and Manila were on csi-snapshotter v1.2.2 (or at least the csi-cinder's Helm chart still is).

**Comment by jichenjc** [comment]
> @jichenjc Ah, I missed #974. Both CSI Cinder and Manila were on csi-snapshotter v1.2.2 (or at least the csi-cinder's Helm chart still is). And CI job for csi-manila tests snapshotting as well.

should I do additional test? I don't have manila env on my hand ..

**Comment by ramineni** [comment]
> @ramineni and other csi-cinder folks: can you please confirm snapshotting works for you with k8s 1.19 and csi-snapshotter v1.2.2?

@gman0 cinder is updated to 2.1.1 for 1.19 . 
Any update on the manila failure? This is blocking cpo release.
/cc @gouthampacha @tombarron

**Comment by gman0** [comment]
@ramineni OpenLab's [install-k8s role](https://github.com/theopenlab/openlab-zuul-jobs/blob/master/roles/install-k8s/tasks/main.yaml) uses k8s master:

```
2020-09-08 06:54:30.295331 | TASK [install-k8s : Install etcd v3.4.13 and kubernetes master]
```

I've just checked out k8s with branch v1.19.0 and csi-snapshotter v1.2.2 works. Then I tried with k8s master:

```
E0909 15:43:35.798659       1 event.go:247] Could not construct reference to: '&v1alpha1.VolumeSnapshot{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:v1.ObjectMeta{Name:"new-nfs-share-snap", GenerateName:"", Namespace:"default", SelfLink:"", UID:"c06ae5a9-7a84-441a-8fe2-b2b3eea805b0", ResourceVersion:"470", Generation:2, CreationTimestamp:v1.Time{Time:time.Time{wall:0x0, ext:63735262998, loc:(*time.Location)(0x238f460)}}, DeletionTimestamp:(*v1.Time)(nil), DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Initializers:(*v1.Initializers)(nil), Finalizers:[]string{"snapshot.storage.kubernetes.io/volumesnapshot-protection"}, ClusterName:"", ManagedFields:[]v1.ManagedFieldsEntry{v1.ManagedFieldsEntry{Manager:"kubectl", Operation:"Update", APIVersion:"snapshot.storage.k8s.io/v1alpha1", Time:(*v1.Time)(0xc000319bc0), Fields:(*v1.Fields)(nil)}, v1.ManagedFieldsEntry{Manager:"csi-snapshotter", Operation:"Update", APIVersion:"snapshot.storage.k8s.io/v1alpha1", Time:(*v1.Time)(0xc000319c20), Fields:(*v1.Fields)(nil)}}}, Spec:v1alpha1.VolumeSnapshotSpec{Source:(*v1.TypedLocalObjectReference)(0xc0002f16e0), SnapshotContentName:"", VolumeSnapshotClassName:(*string)(0xc000698610)}, Status:v1alpha1.VolumeSnapshotStatus{CreationTime:(*v1.Time)(0xc000319c40), RestoreSize:(*resource.Quantity)(0xc000649dc0), ReadyToUse:false, Error:(*v1beta1.VolumeError)(0xc0002f1710)}}' due to: 'selfLink was empty, can't make reference'. Will not report event: 'Warning' 'SnapshotCreationFailed' 'Failed to create snapshot: selfLink was empty, can't make reference'
E0909 15:43:35.798716       1 snapshot_controller.go:370] createSnapshot [create-default/new-nfs-share-snap[c06ae5a9-7a84-441a-8fe2-b2b3eea805b0]]: error occurred in createSnapshotOperation: selfLink was empty, can't make reference
```

...which is the same error as we're getting from the CI. I suspect this might be a bug in the latest Kubernetes.

**Comment by gman0** [comment]
To not block this PR any further, I'll disable all snapshot-related tests in csi-manila's CI job until it's resolved (or we move to csi-snapshotter v2 - whichever comes first).

**Comment by tombarron** [comment]
Thankis for this out, @gman0.  I had a long weekend (holiday combined with a couple days vacation) and was just looking at this issue now and was not sure what was going on.

**Comment by jichenjc** [comment]
/approve 
I think we are ready to merge this now

**Comment by lingxiankong** [comment]
@jichenjc @ramineni @wangxiyuan @gman0 thanks for your help!

**Pull request merged**